File: src/shared/gestures/engine.ts
import type { Baseline } from "../calibrate";
import type { Action, Config } from "../config";
import type { Features } from "../pose";
import { TriggerAxis } from "./base";
import { SkipDetector } from "./skip";

export interface Signals {
  lane: number;
  jump: number;
  duck: number;
}

/**
 * Turns per-frame Features + Baseline into discrete game actions.
 * Lanes are RELATIVE (one sidestep = one move). A recent move keeps you
 * "alive" for `actionGrace` seconds so moving doesn't trip the skip gate.
 */
export class GestureEngine {
  readonly skip: SkipDetector;
  private left: TriggerAxis;
  private right: TriggerAxis;
  private up: TriggerAxis;
  private down: TriggerAxis;
  private restart: TriggerAxis;

  running = false;
  signals: Signals = { lane: 0, jump: 0, duck: 0 };
  private laneRefX: number | null = null;
  private lastActionAt = -1e9;
  private lastJumpAt = -1e9;

  constructor(private cfg: Config) {
    this.skip = new SkipDetector(cfg.kneeLiftThreshold, cfg.skipTimeout);
    this.left = new TriggerAxis("left", cfg.laneThreshold, cfg.laneRelease, cfg.laneCooldown);
    this.right = new TriggerAxis("right", cfg.laneThreshold, cfg.laneRelease, cfg.laneCooldown);
    this.up = new TriggerAxis("up", cfg.jumpThreshold, cfg.jumpRelease, cfg.cooldown);
    this.down = new TriggerAxis("down", cfg.duckThreshold, cfg.duckRelease, cfg.cooldown);
    this.restart = new TriggerAxis(
      "restart",
      cfg.restartThreshold,
      cfg.restartRelease,
      cfg.restartCooldown,
    );
  }

  update(f: Features, baseline: Baseline, now: number): Action[] {
    const sw = baseline.shoulderWidth;
    if (this.laneRefX === null) this.laneRefX = baseline.hipX;

    const dx = (f.hipX - this.laneRefX) / sw; // + right, - left (mirrored view)
    const jump = (baseline.hipY - f.hipY) / sw; // + up
    const duck = (f.shoulderY - baseline.shoulderY) / sw; // + down
    this.signals = { lane: dx, jump, duck };

    const right = this.right.update(dx, now);
    const left = this.left.update(-dx, now);
    if (right || left) this.laneRefX = f.hipX; // re-center after a lane move

    const up = this.up.update(jump, now);
    if (up) this.lastJumpAt = now;

    // Duck only while standing+crouching, never mid-jump.
    const jumping =
      jump > this.cfg.jumpRelease ||
      this.up.state === "triggered" ||
      now - this.lastJumpAt <= this.cfg.jumpDuckLockout;
    const down = this.down.update(jumping ? -1 : duck, now);

    // Restart: raise both hands above your shoulders (works on the game-over screen).
    const handsUp = (baseline.shoulderY - Math.max(f.wristLY, f.wristRY)) / sw;
    const restart = this.restart.update(handsUp, now);

    const actions = [right, left, up, down].filter((a): a is Action => a !== null);
    if (actions.length) this.lastActionAt = now;

    const skipping = this.skip.update(f, baseline, now);
    this.running = skipping || now - this.lastActionAt <= this.cfg.actionGrace;

    const gated = this.cfg.requireRunning && !this.running ? [] : actions;
    // restart always passes the gate (you're likely stopped on game-over)
    return restart ? [...gated, restart] : gated;
  }
}

File: reference/gestures/engine.py
import time

from .base import TriggerAxis
from .skip import SkipDetector


class GestureEngine:
    """Turns per-frame Features + Baseline into discrete game actions.

    Actions: "left", "right", "up" (jump), "down" (duck).
    Running state is a presence gate; when required, actions are suppressed
    unless the player is actively skipping.
    """

    def __init__(self, cfg):
        self.cfg = cfg
        self.skip = SkipDetector(cfg.knee_lift_threshold, cfg.skip_timeout)
        self.left = TriggerAxis("left", cfg.lane_threshold, cfg.lane_release, cfg.lane_cooldown)
        self.right = TriggerAxis("right", cfg.lane_threshold, cfg.lane_release, cfg.lane_cooldown)
        self.up = TriggerAxis("up", cfg.jump_threshold, cfg.jump_release, cfg.cooldown)
        self.down = TriggerAxis("down", cfg.duck_threshold, cfg.duck_release, cfg.cooldown)
        self.running = False
        self.lane_ref_x = None  # moving reference so lanes are relative
        self.last_action_at = -1e9  # for the action grace window
        self._last_jump_at = -1e9   # to lock out duck around jumps
        self.signals = {"lane": 0.0, "jump": 0.0, "duck": 0.0}

    def update(self, f, baseline, now=None):
        now = now if now is not None else time.monotonic()
        sw = baseline.shoulder_width
        if self.lane_ref_x is None:
            self.lane_ref_x = baseline.hip_x

        # Lane is RELATIVE to the current spot: one step = one move, then re-center.
        dx = (f.hip_x - self.lane_ref_x) / sw         # + right, - left (mirrored view)
        jump = (baseline.hip_y - f.hip_y) / sw        # + up (relative to standing)
        duck = (f.shoulder_y - baseline.shoulder_y) / sw  # + down (relative to standing)
        self.signals = {"lane": dx, "jump": jump, "duck": duck}

        # Detect every gesture regardless of the gate.
        right = self.right.update(dx, now)
        left = self.left.update(-dx, now)
        if right or left:
            self.lane_ref_x = f.hip_x  # re-center after a lane move
        up = self.up.update(jump, now)
        if up:
            self._last_jump_at = now

        # Duck only while standing+crouching, never mid-jump. Suppress duck when the
        # body is rising, the jump axis is still latched, or just after a jump.
        jumping = (jump > self.cfg.jump_release
                   or self.up.state == TriggerAxis.TRIGGERED
                   or (now - self._last_jump_at) <= self.cfg.jump_duck_lockout)
        down = self.down.update(-1.0 if jumping else duck, now)
        actions = [a for a in (right, left, up, down) if a]
        if actions:
            self.last_action_at = now  # a move counts as staying alive

        skipping = self.skip.update(f, baseline, now)
        # A recent move overrides the skip gate for a few seconds.
        self.running = skipping or (now - self.last_action_at) <= self.cfg.action_grace

        if self.cfg.require_running and not self.running:
            return []
        return actions
